70_FLIPFluids/main.cpp

```diff
@@ -0,0 +1,2240 @@
+			camInfo.desc = smart_refctd_ptr(cameraBuffer);
+			camInfo.info.buffer = {.offset = 0, .size = cameraBuffer->getSize()};
+			IGPUDescriptorSet::SDescriptorInfo verticesInfo;
+			verticesInfo.desc = smart_refctd_ptr(particleVertexBuffer);
+			verticesInfo.info.buffer = {.offset = 0, .size = particleVertexBuffer->getSize()};
+			IGPUDescriptorSet::SWriteDescriptorSet writes[2] = {
+				{.dstSet = m_renderDs.get(), .binding = 0, .arrayElement = 0, .count = 1, .info = &camInfo},
+				{.dstSet = m_renderDs.get(), .binding = 1, .arrayElement = 0, .count = 1, .info = &verticesInfo},
+			};
+			m_device->updateDescriptorSets(std::span(writes, 2), {});
+		}
+
+		SBlendParams blendParams = {};
+		blendParams.logicOp = ELO_NO_OP;
+		blendParams.blendParams[0u].srcColorFactor = asset::EBF_SRC_ALPHA;
+		blendParams.blendParams[0u].dstColorFactor = asset::EBF_ONE_MINUS_SRC_ALPHA;
+		blendParams.blendParams[0u].colorBlendOp = asset::EBO_ADD;
+		blendParams.blendParams[0u].srcAlphaFactor = asset::EBF_ONE_MINUS_SRC_ALPHA;
+		blendParams.blendParams[0u].dstAlphaFactor = asset::EBF_ZERO;
+		blendParams.blendParams[0u].alphaBlendOp = asset::EBO_ADD;
+		blendParams.blendParams[0u].colorWriteMask = (1u << 0u) | (1u << 1u) | (1u << 2u) | (1u << 3u);
+
+		{
+			IGPUShader::SSpecInfo specInfo[3] = {
+				{.shader = vs.get()},
+				{.shader = fs.get()},
+			};
+
+			const auto pipelineLayout = m_device->createPipelineLayout({}, nullptr, smart_refctd_ptr(descriptorSetLayout1), smart_refctd_ptr(descriptorSetLayout2), nullptr);
+
+			SRasterizationParams rasterizationParams{};
+			rasterizationParams.faceCullingMode = EFCM_NONE;
+			rasterizationParams.depthWriteEnable = true;
+
+			IGPUGraphicsPipeline::SCreationParams params[1] = {};
+			params[0].layout = pipelineLayout.get();
+			params[0].shaders = specInfo;
+			params[0].cached = {
+				.vertexInput = {
+				},
+				.primitiveAssembly = {
+					.primitiveType = E_PRIMITIVE_TOPOLOGY::EPT_TRIANGLE_LIST,
+				},
+				.rasterization = rasterizationParams,
+				.blend = blendParams,
+			};
+			params[0].renderpass = renderpass;
+
+			if (!m_device->createGraphicsPipelines(nullptr, params, &m_graphicsPipeline))
+				return logFail("Graphics pipeline creation failed");
+		}
+
+		return true;
+	}
+
+	void mouseProcess(const nbl::ui::IMouseEventChannel::range_t& events)
+	{
+		for (auto eventIt = events.begin(); eventIt != events.end(); eventIt++)
+		{
+			auto ev = *eventIt;
+
+			// do nothing
+		}
+	}
+
+
+	// in-loop functions
+	void initializeParticles(IGPUCommandBuffer* cmdbuf)
+	{
+		{
+			SMemoryBarrier memBarrier;
+			memBarrier.srcStageMask = PIPELINE_STAGE_FLAGS::ALL_COMMANDS_BITS;
+			memBarrier.srcAccessMask = ACCESS_FLAGS::MEMORY_READ_BITS | ACCESS_FLAGS::MEMORY_WRITE_BITS;;
+			memBarrier.dstStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+			memBarrier.dstAccessMask = ACCESS_FLAGS::SHADER_READ_BITS;
+			cmdbuf->pipelineBarrier(E_DEPENDENCY_FLAGS::EDF_NONE, {.memBarriers = {&memBarrier, 1}});
+        }
```

Comment:
you destination access should also include writing, because you don't want writes to overlap too

btw if you use IUtilities to create your buffer, they submit right away and such pipeline barrier is not needed